File: 1. Array & String/3.StringtoURL/Solution.go
package main

func replaceSpaces(str string, length int) string {
	// return url.PathEscape(S[:length])
	j := len(str)
	b := []byte(str)
	for i := length - 1; i >= 0; i-- {
		if b[i] == ' ' {
			b[j-1] = '0'
			b[j-2] = '2'
			b[j-3] = '%'
			j -= 3
		} else {
			b[j-1] = b[i]
			j--
		}
	}
	return string(b[j:])
}

File: 1. Array & String/3.StringtoURL/Solution.ts
/**
 * @param {string} S
 * @param {number} length
 * @return {string}
 */
var replaceSpaces = function (S: string, length: number) {
    // have a pointer to check from start to end
    var strArr = S.split('');
    var pointer: number = 0;
    while (pointer < length) {
        if (strArr[pointer] === ' ') {
            // *** needs more work here, a little wierd
            // not handling trailing spaces properly
            for (var i = length - 1; i > pointer + 3; i--) {
                strArr[i] = S[i - 2];
            }
            strArr[pointer] = '%';
            strArr[pointer + 1] = '2';
            strArr[pointer + 2] = '0';
        }
        pointer++;
    }
    // if character is a space, move remainder chars by two
    // replace following three chars with '%20'
    return strArr.join('');
};
